Add Notification System API

shivani-sangwan opened the pull request:
Adding Notification and UserNotification Association database layer.

## How to test the changes?
- Tested the relationships and functioning using the unit tests for the above-mentioned database tables. 

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,100 @@
+from galaxy.web import url_for
+
+
+class NotificationManager:
```
This is looking good, but can you remove all id encoding and decoding from the Manager ? I don't think we need this at all for a modern API. This should all just work with integers.

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,100 @@
+        trans.sa_session.add(notification)
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        trans.app.security_agent.set_entity_notification_associations(users=[users], notification=notification)
```
Why did you move this into the security_agent module ? I think that's business logic that belongs in the manager

shivani-sangwan replied:
Thought to follow the similar pipeline, as done for other associations like groups.py for UserGroupAssociation.

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,100 @@
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+    def __init__(self, app: MinimalManagerApp) -> None:
+        self._app = app
```
You should only need the database session here. If a Manager doesn't need app (a [god-object](https://en.wikipedia.org/wiki/God_object), we want to remove its usage whenever possible) it shouldn't receive it.

shivani-sangwan replied:
Should I use MinimalManagerTool instead, it contains a galaxy scoped session.

mvdbeek replied:
No, you don't need app here at all.

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,100 @@
+        trans.app.security_agent.set_entity_notification_associations(users=[users], notification=notification)
+        trans.sa_session.flush()
+        item = group.to_dict(view="element", value_mapper={"id": trans.security.encode_id})
+        item["url"] = url_for("notification", id=encoded_id)
```
We also want to eliminate usage of `url_for`, is there a specific use you have in mind for the url ?

shivani-sangwan replied:
I was thinking that notifications will be posted to a specific /notification/ page for the user.

mvdbeek replied:
Yes, but why do you need the URL ?

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,100 @@
+        trans.sa_session.flush()
+        item = group.to_dict(view="element", value_mapper={"id": trans.security.encode_id})
+        item["url"] = url_for("notification", id=encoded_id)
+        return [item]
```
Any reason you're return a list here ?

shivani-sangwan replied:
Not a specific reason, just thought if we will have group/role objects then we might need it.

mvdbeek replied:
Yes, but why wrap it in a list ?

mvdbeek commented on test/unit/data/model/test_notification.py:
```diff
@@ -0,0 +1,87 @@
+    return notification
+
+def create_user(sa_session: galaxy_scoped_session):
+    random_number = random.randint(0, 1000)
```
I think the next step here is to use the Manager in these tests ... you're essentially testing that you're using sqlalchemy correctly currently. That's OK, but ideally the unit tests should test our own classes and methods.

jdavcs replied:
we have infrastructure for that - see my separate comment on testing the mapping.

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        """
+        rval = []
+        for notification in trans.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
+            item = notification.to_dict(value_mapper={"id": notification.id})
```
```suggestion
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        rval = []
+        for notification in trans.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
+            item = notification.to_dict(value_mapper={"id": notification.id})
+            rval.append(item)
```
```suggestion
            rval.append(notification)
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+            rval.append(item)
+        return rval
+
+    def create(self, trans: ProvidesAppContext, payload: Dict[str, Any]):
```
```suggestion
    def create(self, trans: ProvidesAppContext, message: str):
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        if message_text is None:
+            raise ObjectAttributeMissingException("Missing required message")
+
+        notification = model.Notification(message_text=message_text)
```
```suggestion
        notification = model.Notification(message_text=message)
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        trans.sa_session.flush()
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        self._associate_user_notification(users=[users], notification=notification)
```
```suggestion
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        """
+        notification = self._get_notification(trans, user_id)
+        item = notification.to_dict(view="element", value_mapper={"id": notification.id})
+        return item
```
```suggestion
    def get(self, notification_id) -> model.Notification:
        """
        Get a single notification
        """
        return = self._get_notification(trans, notification_id)
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+
+
+class NotificationManager:
+    """Interface/service object shared by controllers for interacting with notifications."""
+
```
```suggestion
    """Interface/service object shared by controllers for interacting with notifications."""

    def __init__(self, sa_session: galaxy_scoped_session):
        self.sa_session = sa_session

```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        Displays a collection (list) of notifications.
+        """
+        rval = []
+        for notification in trans.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
```
```suggestion
        for notification in self.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+
+    def index(self, trans: ProvidesAppContext):
```
```suggestion
    def index(self, limit: int = 10, offset: int = 0, user: Optional[model.User] = None):
```

mvdbeek replied:
I'll leave the implementation of filtering for users, limit and offset as an exercise for you to implement

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        trans.sa_session.flush()
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        self._associate_user_notification(users=[users], notification=notification)
```
```suggestion
    def update(self, notification: model.Notification, message: str):
        """
        Modifies a notification.
        """
        notification.message_text = message_text
        trans.sa_session.add(notification)
        trans.sa_session.flush()
        return notification
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        notification = trans.sa_session.query(model.Notification).get(user_id)
+        if notification is None:
+            raise ObjectNotFound(f"Notification with id {user_id} was not found.")
+        return notification
```
```suggestion
    def _get_notification(self, user_id: int) -> model.Notification:
        notification = self.sa_session.query(model.Notification).get(user_id)
        if notification is None:
            raise ObjectNotFound(f"Notification with id {user_id} was not found.")
        return notification
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        decoded_user_ids = self._decode_ids(encoded_user_ids)
+        users = trans.sa_session.query(model.User).filter(model.User.table.c.id.in_(decoded_user_ids)).all()
+        return users
+
```
```suggestion
```

That's what the user manager does, that doesn't need to go here.

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+        for user in users:
+            assoc = model.UserNotificationAssociation(user, notification)
+            trans.sa_session.add(assoc)
+            trans.sa_session.flush()
```
```suggestion
    def _associate_user_notification(self, users: List[model.user], notification: model.Notification):
        for user in users:
            assoc = model.UserNotificationAssociation(user, notification)
            trans.sa_session.add(assoc)
        trans.sa_session.flush()
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,94 @@
+from galaxy.managers.context import ProvidesAppContext
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.structured_app import MinimalToolApp
+from galaxy.web import url_for
```
```suggestion
```

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,61 @@
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+    def __init__(self, sa_session: galaxy_scoped_session):
+        self.sa_session = galaxy_scoped_session
```
```suggestion
        self.sa_session = sa_session
```
this is why your tests are failing

shivani-sangwan committed:
Add Notification System API Database with migration

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -2409,6 +2409,42 @@ def __init__(self, user, group):
+    create_time = Column(DateTime, default=now)
+    update_time = Column(DateTime, default=now, onupdate=now)
+    user = relationship("User", backref="all_notifications")
+    notification = relationship("Notification", backref="user_notification_associations")
```
Please use `back_populates` instead of `backref` (you'll also need to add the opposite relationship to the `Notification` class. (see #13165 for context)

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -8127,7 +8163,6 @@ def _serialize(self, id_encoder, serialization_options):
         request_input_collection_attrs["order_index"] = self.workflow_step.order_index
         return request_input_collection_attrs
 
-
```
2 spaces between class definitions is correct

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -8176,7 +8211,6 @@ def _serialize(self, id_encoder, serialization_options):
         output_dataset_attrs["workflow_output"] = self.workflow_output.serialize(id_encoder, serialization_options)
         return output_dataset_attrs
 
-
```
same as above

shivani-sangwan committed:
Adding mapping tests for Notification and UserNotificationAssociation

shivani-sangwan committed:
Adding fxitures and mapper tests for Notification and UserNotification

shivani-sangwan committed:
Adding API module for Notification Manager

shivani-sangwan committed:
Adding notification module in API and registering NotificationManager in structured_app.py

shivani-sangwan committed:
Adding API tests for Notifications

mvdbeek committed:
Move notiifcation api tests to lib/galaxy_test/api

shivani-sangwan committed:
Adding Payload Descriptor for Notification in Schema

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -3043,6 +3057,14 @@ class Config:
+    
+    class Config:
+        use_enum_values = True  # When using .dict()
+        extra = Extra.allow  # Allow any other extra fields
```
why did you add this?

mvdbeek commented on lib/galaxy/managers/notification.py:
```diff
@@ -0,0 +1,75 @@
+            rval.append(notification)
+        return rval
+
+    def create(self, trans, payload):
```
You do not need trans here, you can use self.sa_session. You should also add the type annotation for payload here.

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -3003,6 +3004,19 @@ class PageSummaryBase(BaseModel):
+)
+
+class NotificationBase(BaseModel):
+    message: str = Field(
```
This should be message_text to be consistent with the database model.
You also need a response model with the properties you want to return. That can be the same model for now.

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -3043,6 +3057,14 @@ class Config:
 
 
+class CreateNotificationPayload(NotificationBase):
+    notification: Optional[str] = NotificationContentField
```
I think you're overcomplicating things here. All you need in the CreateNotificationPayload is the message_text property, i don't think this requires separately defined fields, just do define it here.

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -3043,6 +3057,14 @@ class Config:
+    notification: Optional[str] = NotificationContentField
+    
+    class Config:
+        use_enum_values = True  # When using .dict()
```
You're not using any enums

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -46,6 +46,7 @@
 
 USER_MODEL_CLASS_NAME = "User"
 GROUP_MODEL_CLASS_NAME = "Group"
+NOTIFICATION_MODEL_CLASS_NAME = "Notification"
```
you're not using this

mvdbeek commented on lib/galaxy/webapps/galaxy/api/notifications.py:
```diff
@@ -0,0 +1,51 @@
+    def show(self, notification_id):
+        return self.manager.show(notification_id)
+
+    @router.post("/api/notifications", summary="Create a notificaton message")
```
You need to annotate a response model here, you can't just return the database model here. This is true for all the methods here.

shivani-sangwan committed:
Adding notification annotation response in API

mvdbeek commented on lib/galaxy/webapps/galaxy/api/notifications.py:
```diff
@@ -0,0 +1,65 @@
+    def show(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        notification_id: NotificationIdField = NotificationIdParam,
```
```suggestion
        notification_id: NotificationIdParam,
        trans: ProvidesAppContext = DependsOnTrans,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/notifications.py:
```diff
@@ -0,0 +1,65 @@
+    def create(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        message_text: NotificationMessageField = NotificationMessageParam,
```
```suggestion
        message_text: NotificationMessageParam,
        trans: ProvidesAppContext = DependsOnTrans,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/notifications.py:
```diff
@@ -0,0 +1,65 @@
+router = Router(tags=["notifications"])
+
+NotificationIdParam: NotificationIdField = Path(..., title="ID", description="Notification ID of the notification")
+
```
```suggestion
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/notifications.py:
```diff
@@ -0,0 +1,65 @@
+NotificationIdParam: NotificationIdField = Path(..., title="ID", description="Notification ID of the notification")
+
+NotificationMessageParam: NotificationMessageField = Path(..., title="Message", description="Message of the notification")
+
```
```suggestion


```

mvdbeek committed:
Remove unique constraint on notification message text

mvdbeek committed:
Fix offset logic in NotificationManager index

mvdbeek committed:
Fix usage of schema models

shivani-sangwan committed:
Adding NotificationRequestModel in schema

shivani-sangwan committed:
Adding unit tests for notifications API

mvdbeek commented on lib/galaxy_test/api/test_notifications.py:
```diff
@@ -0,0 +1,50 @@
+        payload = {"message_text": "Test Message"}
+        notification_response = self._post("notifications", payload, json=True)
+        notification_response.raise_for_status()
+        payload = {"limit": 5, "offset": 0}
```
The types of your notification dictionary are changing, that's why you're getting:
```
lib/galaxy_test/api/test_notifications.py:14: error: Dict entry 0 has
incompatible type "str": "int"; expected "str": "str"  [dict-item]
            payload = {"limit": 5, "offset": 0}
```

mvdbeek commented on lib/galaxy_test/api/test_notifications.py:
```diff
@@ -0,0 +1,50 @@
+        notification_response = self._post("notifications", payload, json=True)
+        notification_response.raise_for_status()
+        payload = {"limit": 5, "offset": 0}
+        notification_response = self._get("notifications", payload)
```
```suggestion
        create_notification_payload = {"message_text": "Test Message"}
        notification_response = self._post("notifications", create_notification_payload, json=True)
        notification_response.raise_for_status()
        notification_list_parameters = {"limit": 5, "offset": 0}
        notification_response = self._get("notifications", notification_list_parameters)
```

shivani-sangwan committed:
Adding UI component for notification display

commit 2a4e8575980f8bb736c87d298ca89b33dce2b8f2
Author: Shivani Sangwan <shivanisangwan31@gmail.com>
Date:   Tue Aug 9 19:32:34 2022 +0530

    Adding UI for notification list display

dannon commented on client/src/components/Notification/List.vue:
```diff
@@ -0,0 +1,59 @@
+        <loading-span v-if="loading" message="Loading notifications" />
+        <ul v-else id="notifications">
+            <li v-for="notifiation in notifications" :key="notifications" :data-extension-target="notifications">
+                <span
```
```suggestion
                <span>
```
This span open tag is missing its right arm.

dannon commented on client/src/components/Notification/List.vue:
```diff
@@ -0,0 +1,59 @@
+                        <i class="close icon"></i>
+                        <div class="header">
+                           {{ notifiation.message_text }}
+                        </div>
```
```suggestion
                        </div>
                    </div>
```

Missing a closing div here.

shivani-sangwan committed:
Adding user notification association while creating notifications
